### PR TITLE
fix e2e producer

### PR DESCRIPTION
**Pull request by bachmanity1**
image

I've noticed that when working with clusters having several controller and broker nodes, retrieving cluster metadata may occasionally return stale data. This issue is evident when, for example, an end-to-end test topic is created, and the topic metadata is retrieved immediately afterward. In this case, the metadata request may return error code 3, which corresponds to the `UNKNOWN_TOPIC_OR_PARTITION` error. This error likely occurs when a client connects to a broker that is not yet aware of the new topic because it hasn't consumed the latest data from the `__cluster_metadata` topic.

In summary, when the end-to-end code attempts to populate the value of `s.partitionCount`, it might encounter the `UNKNOWN_TOPIC_OR_PARTITION` error, even if the topic actually exists. This results in a partition count of zero, preventing the producer from sending any data.

### CODE REVIEW

File: e2e/service.go [hunk 1]
```diff
@@ -155,14 +155,6 @@ func (s *Service) Start(ctx context.Context) error {
 		return fmt.Errorf("could not validate end-to-end topic: %w", err)
 	}
 
-	// Get up-to-date metadata and inform our custom partitioner about the partition count
-	topicMetadata, err := s.getTopicMetadata(ctx)
-	if err != nil {
-		return fmt.Errorf("could not get topic metadata after validation: %w", err)
-	}
-	partitions := len(topicMetadata.Topics[0].Partitions)
```
Comment (bachmanity1) [inline e2e/service.go:163]: `topicMetadataTopics[0].ErrorCode` is not guaranteed to be zero.

**Comment by bachmanity1** [comment]
Hi @weeco. Can you please have a look?

**Comment by bachmanity1** [comment]
@weeco just a kind reminder.

**Comment by bachmanity1** [comment]
@andrewhsu can you have a look?

**Comment by weeco** [comment]
@bachmanity1 I'll try to look at your PR soon, but I'm currently very busy with other things that have higher priority. Even though this change is just a few lines, I'll have to dig deeper to see if this could cause other sort of issues before approving & merging the PR.

**Comment by bachmanity1** [comment]
Hi @weeco,

I wanted to let you know that I've updated this PR. I realized that my previous approach didn't account for scenarios where the broker count is updated.

I also wanted to discuss the `partitionsPerBroker` config. I believe it should clarified that this configuration is only applied when a topic is created and does not take effect when the broker count changes. If the original intent was for this setting to apply even when the broker count changes, we’d need to adjust the topic reconciliation logic accordingly. However, handling cases where the broker count decreases could be quite tricky.

By the way, it seems there’s a bug preventing new partitions from being created when new brokers are added, as reported in #257. Could you take a look at that PR too? Thanks!

**Comment by bachmanity1** [comment]
@weeco kind reminder

**Comment by bachmanity1** [comment]
Hi @weeco, is there any chance to get this reviewed anytime soon?

**Comment by weeco** [comment]
Thanks for your PR @bachmanity1 and sorry for the long waiting time. I added some minor things but generally LGTM

**Review by weeco — APPROVED** [review]
